### ISSUE TITLE
Inconsistent handling between Str.words and «....» quoters

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#1562]
Str.words just splits on `\s+`, while the routine that breaks up `«...»` [special-cases 4 non-break chars](https://github.com/perl6/nqp/blob/b6d98eac3f0dff1dbd4df252351efa35ab700f7e/src/HLL/Grammar.nqp#L349-L354) (although one of them isn't whitespace and wouldn't be split on anyway).

```
<Zoffix__> m: dd «a\c[NO-BREAK SPACE]bc d»
<camelia> rakudo-moar 7d72387bc: OUTPUT: «("a bc", "d")␤»
<Zoffix__> m: dd "a\c[NO-BREAK SPACE]bc d".words
<camelia> rakudo-moar 7d72387bc: OUTPUT: «("a", "bc", "d").Seq␤»
```

There was a bunch of discussion with no clear agreement: https://irclog.perlgeek.de/perl6/2018-02-24#i_15854872

Personally, I'm against special-cases on hardcoded chars.

**Comment by tbrowder** [comment]
My vote is to treat a non-breaking space as the user obviously intended, i.e., treat the words on either side of it, along with the space, as one word, at least as the default behavior.  My use of .words would not conflict with that, and that is what I would expect.  Check out Professor Donald Knuth’s handling of whitespace and non-breaking whitespace and I think you will see that he handles it the same.

However, maybe another variant of .words, something like “.words-all” could split on *all* whitespace.

**Comment by zoffixznet** [comment]
> as the user obviously intended

What exactly makes it so obvious? I think the PRO crowd is conflating presentation with semantic meaning. [Unicode Line Break docs](http://unicode.org/reports/tr14/#Conformance) for example consistently talk about **layout** when it comes to breaks.

Sure, `"Zoffix\x[A0]Znet"` shouldn't be broken across separate lines when rendered, but it's still two words, not one. Just as `"zoffix\x[A0]znet".wordcase` returns `"Zoffix\x[A0]Znet"` not `"Zoffix\x[A0]znet"`, so should `"Zoffix\x[A0]Znet".words.perl` return `("Zoffix", "Znet").Seq` not `("Zoffix\x[0]Znet").Seq`.


I see the proposal to special case several arbitrary characters as a means to DWIM, yet to me that DWIM seems to assume all data will be always perfectly formatted. The Web is littered with pages that use non-breaking space as a spacer. [`DOM::Tiny`](https://modules.perl6.org/repo/DOM::Tiny) goes as far as converting all `&nbsp;`s to regular spaces to avoid problems. And if you've ever used [`Mojo::DOM`](https://metacpan.org/pod/Mojo::DOM)—which doesn't do such treatment—you'd be quite familiar with just how prevalent those spacers are.

Trying to remember the codepoints of special-cased characters adds to the learning curve and is an extra cognitive load to the users. I posit that instead of punishing all users with that extra load, we let the users who can guarantee their data is well-formatted and who are aware of these special characters to merely write the slightly-longer, *more explicit, and self-documenting* `.split: /<[\s]-\x[A0] ... >/`.

Considering `.words` and `.wordcase` already behave the way I propose they do, from the backwards-compatibility viewpoint, I maintain keeping them that way is a superior alternative.

----- 

> Donald Knuth

He also said it's too late for Perl 6. I think I'll ignore his opinions.

**Comment by tbrowder** [comment]
To be clear, I was only thinking of single spaces between words, not multiple non-breaking spaces.  So that tends to support your position, that non-breaking spaces should not be treated differently. Although for my use cases, I would normalize contiguous non-breaking spaces to a single one.  Granted, that again tends to support your view.  Frankly, I hadn’t really thought of the problem since I normally don’t think unicode or html when I think of text handling.

**Comment by AlexDaniel** [comment]
I think it's more complicated, and IMO .words should be doing some fancy unicode-style splitting (anyone who doesn't want to use that can always do `.comb(/\w+/)`). Maybe see https://www.unicode.org/reports/tr29/#Word_Boundaries and https://unicode.org/cldr/utility/breaks.jsp. But if that's implemented, the functionality is going to change *a lot* (like ‘foo-bar’ will be two words instead of one).

:man_shrugging:

**Comment by AlexDaniel** [comment]
Oh, as for `« »`, I suggest we kill it with fire.

**Comment by AlexDaniel** [comment]
Ah, a little mistake on my part. The behavior seems to be the same in « » and < > (the latter of course shouldn't be killed), so I guess we'd still need to discuss that :)

**Comment by zoffixznet** [comment]
> it's more complicated

No, it isn't. We have a simple routine that does a simple, easy-to-understand thing and no one complained about it or misunderstood its behaviour in 2 years. There's no reason to over-engineer it and [disrupt our users who use it](https://gist.github.com/Whateverable/234040882f4ae2317d2e256969d4a08c).

There isn't a common use where you'd want `"Sample Text."` to give you a 3-element list, last of which is a period (or 2-element list with part of actual data discarded). That's a job for a `Lingua::` module, not a Hoffmanized core method call. The common use is splitting on `/\s+/` which in Perl 5 was special-cased as `split ' '` and in Perl 6 has a concise name for it: `.words`. You are ignoring that common use case and wish to invent a feature that best fits the method name instead. Literally going backwards: a feature from a name, rather than a name for a feature.

> can always do .comb(/\w+/)

No, you can't. Doing that on an IO::Handle will slurp everything, while `.words` will do it lazily. And if `.words` is changing its definition of what a word is, `.wordcase` would also have to change.

> Oh, as for « », I suggest we kill it with fire.

Yeah, and why? Because it *is* over-engineered.

------

There's a joke I saw on Twitter: <br>*Perl 6 is what happens when every bikeshedding session ends with "LET'S DO BOTH!"*

I see @tbrowder already proposed exactly that above...

**Comment by AlexDaniel** [comment]
> You are ignoring that common use case and wish to invent a feature that best fits the method name instead. Literally going backwards: a feature from a name, rather than a name for a feature.

Okay, fair enough, I guess you're right.

> Doing that on an IO::Handle will slurp everything, while .words will do it lazily

That sounds like a NYI in .comb and .split which I've already struggled with previously. Something about it in [RT#131923](https://rt.perl.org/Ticket/Display.html?id=131923).



> > Oh, as for « », I suggest we kill it with fire.
> 
> Yeah, and why? Because it is over-engineered.

For me primarily because [“Word splitting is a broken legacy misfeature inherited from the Bourne shell that's stuck on by default if you don't quote expansions.”](http://mywiki.wooledge.org/BashPitfalls). How that managed to move into a modern language is completely beyond me. FWIW here are some examples of «» being used in modules: https://gist.github.com/576bc17fe1eb2c081819a4e13e4a7f26.

**Comment by jstuder-gh** [comment]
I was on team "NO-BREAK", but I think you are right zoffix that poorly formatted source data (whether or not .words is considered the right tool for the job, it might be what some users reach for) and backwards compatibility is a real concern.

At the risk of further "Let's do both"ing the bikeshed, what about an adverb to .words that will enable the "NO-BREAK" behavior. This way users can declaratively signal intent and backwards compatibility is preserved.

    say "a\c[NO-BREAK SPACE]bc d".words\      .perl 
    # OUTPUT: «("a", "bc", "d").Seq␤»

    say "a\c[NO-BREAK SPACE]bc d".words(:nbsp).perl
    # OUTPUT: «("a bc", "d").Seq␤»

TimToady said this in the linked IRC exchange that resonated with me:

> I just know I get really ticked off when a program thinks it knows better than me whether to
> split up "Perl 6"
https://irclog.perlgeek.de/perl6/2018-02-24#i_15855043

I still think that the "NO-BREAK" behavior should be there in some capacity.
